external-crates/move/crates/move-analyzer/trace-adapter/src/source_map_utils.ts

```diff
@@ -0,0 +1,174 @@
+    enum_map: Record<string, any>;
+    function_map: Record<string, ISrcFunctionMapEntry>;
+    constant_map: Record<string, any>;
+}
+
+// Runtime data types.
+
+/**
+ * Describes a location in the source file.
+ */
+interface ILoc {
+    line: number;
+    column: number;
+}
+
+/**
+ * Describes a function in the source map.
+ */
+interface ISourceMapFunction {
+    // Locations indexed with PC values.
+    pcLocs: ILoc[]
+}
+
+/**
+ * Information about a Move source file.
+ */
+export interface IFileInfo {
+    // File path.
+    path: string;
+    // File content.
+    content: string;
+    // File content split into lines (for efficient line/column calculations).
+    lines: string[];
+}
+
+/**
+ * Source map for a Move module.
+ */
+export interface ISourceMap {
+    fileHash: string
+    modInfo: ModuleInfo,
+    functions: Map<string, ISourceMapFunction>
+}
+
+export function readAllSourceMaps(directory: string, filesMap: Map<string, IFileInfo>): Map<string, ISourceMap> {
+    const sourceMapsMap = new Map<string, ISourceMap>();
+
+    const processDirectory = (dir: string) => {
+        const files = fs.readdirSync(dir);
+        for (const f of files) {
+            const filePath = path.join(dir, f);
+            const stats = fs.statSync(filePath);
+            if (stats.isDirectory()) {
+                processDirectory(filePath);
+            } else if (path.extname(f) === ".json") {
+                const sourceMap = readSourceMap(filePath, filesMap);
+                sourceMapsMap.set(JSON.stringify(sourceMap.modInfo), sourceMap);
+            }
+        }
+    };
+
+    processDirectory(directory);
+
+    return sourceMapsMap;
+}
+
+/**
+ * Reads a Move VM source map from a JSON file.
+ * 
+ * @param sourceMapPath path to the source map JSON file.
+ * @param filesMap map from file hash to file information.
+ * @returns source map.
+ */
+function readSourceMap(sourceMapPath: string, filesMap: Map<string, IFileInfo>): ISourceMap {
+    const sourceMapJSON: ISrcRootObject = JSON.parse(fs.readFileSync(sourceMapPath, 'utf8'));
+
+    const fileHash = Buffer.from(sourceMapJSON.definition_location.file_hash).toString('base64');
+    const modInfo: ModuleInfo = {
+        addr: sourceMapJSON.module_name[0],
+        name: sourceMapJSON.module_name[1]
+    };
+    const functions = new Map<string, ISourceMapFunction>();
+    const fileInfo = filesMap.get(fileHash);
+    if (!fileInfo) {
+        throw new Error("Could not find file with hash: "
+            + fileHash
+            + " when processing source map at: "
+            + sourceMapPath);
+    }
+    for (const funEntry of Object.values(sourceMapJSON.function_map)) {
+        let nameStart = funEntry.definition_location.start;
+        let nameEnd = funEntry.definition_location.end;
+        const funName = fileInfo.content.slice(nameStart, nameEnd);
+        const pcLocs: ILoc[] = [];
+        let prevPC = 0;
+        let prevLoc = { line: 0, column: 0 };
```

Comment:
line/col numbering here, and below in `byteOffsetToLineColumn` seem to make different assumptions about if they are 1- or 0-based.

Reply:
Great question and my bad for not explaining more thoroughly. This `prevLoc` value is "fake" (as there is no real "initial" value of the location where the debugger should start). It needs to be initialized to make the compiler happy, but it will never be used: first PC in the frame is always 0 so inner loop (where`prevLoc` is used) never executes on the first iteration of the outer loop. I added a comment explaining this and changed line/column to -1 to avoid confusion about 0/1 based lines/columns

Reply:
Thanks for the explanation!